### PR TITLE
Added configuration parameter to limit the mem capacity of query.

### PR DESCRIPTION
**Pull request by OfirMos**

### CODE REVIEW

File: tests/flow/test_config.py [hunk 1]
```diff
@@ -54,6 +60,19 @@ def test03_config_set(self):
         expected_response = [config_name, config_value]
         self.env.assertEqual(response, expected_response)
 
+
+        config_name = "QUERY_MEM_CAPACITY"
+        config_value = 1<<30 #1GB
```
Comment (swilly22) [inline tests/flow/test_config.py:65]: `# 1GB` space between `#` and the content of the comment

**Review by swilly22 — CHANGES_REQUESTED** [review]
nice work, see comments.

File: src/module.c [hunk 2]
```diff
@@ -38,6 +38,7 @@
 // Module-level global variables
 //------------------------------------------------------------------------------
 RG_Config config;                   // Module global configuration.
+void (*on_config_change)(Config_Option_Field type, void *val); // callback function which called when config param changed
```
Comment (swilly22) [inline src/module.c:41]: Define a type for this function signature,
the declaration of this type should be in `config.h`

File: src/config.h [hunk 3]
```diff
@@ -67,3 +67,5 @@ bool Config_Option_set(Config_Option_Field field, RedisModuleString *val);
 
 bool Config_Option_get(Config_Option_Field field, ...);
 
+// Sets the callback function which should be called when a config param changes.
+void Config_Subscribe_Changes(void (*callback)(Config_Option_Field type, void *val));
```
Comment (swilly22) [inline src/config.h:71]: `void Config_Subscribe_Changes(function_type cb);`
set config.cb to input: `config.cb = cb;`

all files should end with newline, usually the last function in a file is the Free function.

File: src/config.c [hunk 4]
```diff
@@ -54,6 +54,8 @@
 #define VKEY_MAX_ENTITY_COUNT_DEFAULT 100000
 
 extern RG_Config config; // global module configuration
+extern void (*on_config_change)(Config_Option_Field type, void *val); // callback function which called when config param changed
+#define _on_config_change(type, val) if(on_config_change) { on_config_change(type, val); }
```
Comment (swilly22) [inline src/config.c:58]: remove

File: src/config.c [hunk 5]
```diff
@@ -521,6 +532,7 @@ bool Config_Option_set(Config_Option_Field field, RedisModuleString *val) {
 				if (!_Config_ParseInteger(val, &query_mem_capacity)) return false;
 
 				Config_query_mem_capacity_set(query_mem_capacity);
+				_on_config_change(field, &query_mem_capacity);
```
Comment (swilly22) [inline src/config.c:535]: Remove all calls to `_on_config_change`
call `_on_config_change` just before returning `true`

File: src/config.c [hunk 6]
```diff
@@ -590,115 +602,120 @@ bool Config_Option_get(Config_Option_Field field, ...) {
...
+		return false;
+	}
 
 	return true;
 }
+
+void Config_Subscribe_Changes(void (*callback)(Config_Option_Field type, void *val))
+{
+	on_config_change = callback;
+}
```
Comment (swilly22) [inline src/config.c:721]: save callback function within the config object.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/config.h [hunk 7]
```diff
@@ -63,3 +70,6 @@ bool Config_Option_set(Config_Option_Field field, RedisModuleString *val);
 
 bool Config_Option_get(Config_Option_Field field, ...);
 
+// Sets the callback function which should be called when a config param changes.
```
Comment (swilly22) [inline src/config.h:73]: "which should be called.." -> "which will be called..."

File: src/config.c [hunk 8]
```diff
@@ -481,13 +511,29 @@ bool Config_Option_set(Config_Option_Field field, RedisModuleString *val) {
...
+	//----------------------------------------------------------------------
+	// invalid option
+	//----------------------------------------------------------------------
+
+	default:
+		return false;
+	}
+
+	if(config.cb)
+		config.cb(field);
```
Comment (swilly22) [inline src/config.c:536]: one liner: `if(config.cb) config.cb(field);`

File: src/config.c [hunk 9]
```diff
@@ -644,3 +705,8 @@ bool Config_Option_get(Config_Option_Field field, ...) {
 
 	return true;
 }
+
+void Config_Subscribe_Changes(Config_on_change cb) {
+	config.cb = cb;
```
Comment (swilly22) [inline src/config.c:710]: Check `cb` isn't NULL
 `ASSERT(cb != NULL);`

**Review by swilly22 — CHANGES_REQUESTED** [review]
a few small comments

File: src/config.c [hunk 10]
```diff
@@ -378,7 +396,7 @@ int Config_Init(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 		}
 
 		// exit if encountered an error when setting configuration
-		if(!Config_Option_set(field, val)) {
+		if(!Config_Option_set_from_Redis_String(field, val)) {
```
Comment (swilly22) [inline src/config.c:399]: switch back to `Config_Option_set`

File: src/config.c [hunk 11]
```diff
@@ -388,7 +406,14 @@ int Config_Init(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 	return REDISMODULE_OK;
 }
 
-bool Config_Option_set(Config_Option_Field field, RedisModuleString *val) {
+bool Config_Option_set_from_Redis_String(Config_Option_Field field, RedisModuleString *val) {
+	size_t len;
+	const char *val_cstr = RedisModule_StringPtrLen(val, &len);
+	if(!len) return false; // Return err code if string len is 0
+	return Config_Option_set(field, val_cstr);
+}
```
Comment (swilly22) [inline src/config.c:414]: Remove

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/unit/test_thread_pools.cpp [hunk 12]
```diff
@@ -29,6 +30,7 @@ class ThreadPoolsTest: public ::testing::Test {
 	static void SetUpTestCase() {
 		Alloc_Reset();
 		Config_Init(NULL, NULL, 0);
+		Config_Subscribe_Changes(&reconf_handler);
```
Comment (swilly22) [inline tests/unit/test_thread_pools.cpp:33]: Is this call really necessary here?

File: src/util/thpool/thpool.c [hunk 13]
```diff
@@ -56,11 +58,12 @@ typedef struct job {
...
-	job *front;              /* pointer to front of queue */
-	job *rear;               /* pointer to rear  of queue */
-	bsem *has_jobs;          /* flag as binary semaphore  */
-	int len;                 /* number of jobs in queue   */
+	pthread_mutex_t rwmutex; 		/* used for queue r/w access */
+	job *front;              		/* pointer to front of queue */
+	job *rear;               		/* pointer to rear  of queue */
+	bsem *has_jobs;          		/* flag as binary semaphore  */
+	int len;                 		/* number of jobs in queue   */
+	uint64_t max_queued_queries;		/* limit on number of queries */
```
Comment (swilly22) [inline src/util/thpool/thpool.c:66]: rename from `max_queued_queries` to `cap` this is a generic thread pool (not necessarily running queries)

File: src/util/thpool/thpool.h [hunk 14]
```diff
@@ -205,12 +205,20 @@ int thpool_num_threads(threadpool);
...
-uint thpool_queue_size(threadpool);
+bool thpool_queue_full(threadpool);
+
+/**
+ * @brief Sets the limit on max queued queries in pool's jobqueue.
+ *
+ * @param threadpool    the threadpool of interest
+ * @param uint64_t      max queued queries
+ */
+void thpool_set_max_queued_queries(threadpool, uint64_t);
```
Comment (swilly22) [inline src/util/thpool/thpool.h:221]: rename to `thpool_set_jobqueue_cap` update comments accordingly

File: src/util/thpool/thpool.c [hunk 15]
```diff
@@ -272,8 +275,16 @@ int thpool_get_thread_id(thpool_* thpool_p, pthread_t pthread) {
...
+// return true if thread pool internal queue is full with pending work
+bool thpool_queue_full(thpool_* thpool_p) {
+	ASSERT(thpool_p != NULL);
+
+	// test if there's enough room in thread pool queue
+	return (thpool_p->jobqueue.len >= thpool_p->jobqueue.max_queued_queries);
+}
+
+void thpool_set_max_queued_queries(thpool_* thpool_p, uint64_t val) {
+	thpool_p->jobqueue.max_queued_queries = val;
```
Comment (swilly22) [inline src/util/thpool/thpool.c:287]: Assert that `thpool_p` isn't `NULL`

File: src/util/thpool/thpool.c [hunk 16]
```diff
@@ -9,6 +9,8 @@
    ********************************/
 
 //#define _POSIX_C_SOURCE 200809L
+#include "../../config.h"
```
Comment (swilly22) [inline src/util/thpool/thpool.c:12]: I don't believe we need this include

File: src/util/thpool/thpool.c [hunk 17]
```diff
@@ -406,6 +417,10 @@ static int jobqueue_init(jobqueue *jobqueue_p) {
 	pthread_mutex_init(&(jobqueue_p->rwmutex), NULL);
 	bsem_init(jobqueue_p->has_jobs, 0);
 
+	if(!Config_Option_get(Config_MAX_QUEUED_QUERIES, &jobqueue_p->max_queued_queries)) {
+		return -1;
+	}
+
```
Comment (swilly22) [inline src/util/thpool/thpool.c:423]: Replace this with `UINT64_MAX` to indicate unlimited queue size
`jobqueue_p->cap = UINT64_MAX`;

File: src/util/thpool/pools.h [hunk 18]
```diff
@@ -65,3 +65,6 @@ int ThreadPools_AddWorkBulkLoader
 	void *arg_p
 );
 
+// Sets the limit on max queued queries in the jobqueue of all pools.
```
Comment (swilly22) [inline src/util/thpool/pools.h:68]: Rephrase `Sets the limit on max queued queries in each thread pool`

File: src/util/thpool/pools.h [hunk 19]
```diff
@@ -65,3 +65,6 @@ int ThreadPools_AddWorkBulkLoader
...
+void ThreadPools_Set_max_queued_queries(uint64_t val);
```
Comment (swilly22) [inline src/util/thpool/pools.h:69]: consider renaming, `Pools_Set_max_pending_work`

File: src/module.c [hunk 20]
```diff
@@ -60,6 +60,48 @@ static void _PrepareModuleGlobals(RedisModuleCtx *ctx, RedisModuleString **argv,
 	process_is_child = false;
 }
 
+// handler function which being called when config param changed
+bool reconf_handler(Config_Option_Field type) {
```
Comment (swilly22) [inline src/module.c:64]: this function should be placed in a new file, e.g. `reconf_handler.c`

File: src/config.h [hunk 21]
```diff
@@ -27,7 +28,7 @@ typedef enum {
 	Config_END_MARKER               = 10
 } Config_Option_Field;
 
-typedef void (*Config_on_change)(Config_Option_Field type); // callback function which being called when config param changed
+typedef bool (*Config_on_change)(Config_Option_Field type); // callback function which being called when config param changed
```
Comment (swilly22) [inline src/config.h:31]: why the change from `void` to `bool` ?

File: src/config.c [hunk 22]
```diff
@@ -557,7 +551,7 @@ bool Config_Option_set(Config_Option_Field field, const char *val) {
 		return false;
 	}
 
-	if(config.cb) config.cb(field);
+	if(config.cb) return config.cb(field);
 
 	return true;
```
Comment (swilly22) [inline src/config.c:556]:
```
if(config.cb) config.cb(field);
return true;
```
The new value is already set...
the only way i would return the value from the callback is if i'm supporting rollback of to the previous configuration

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/util/thpool/thpool.c (hunk 16 above)
Comment (OfirMos) [reply src/util/thpool/thpool.c:12]: I think we need it because in jobqueue_init we call Config_Option_get

File: src/config.h (hunk 21 above)
Comment (OfirMos) [reply src/config.h:31]: Because the handler calling 				if(!Config_Option_get(type, &max_queued_queries)) {
which might return error and I thought it should be propagated.

File: src/util/thpool/thpool.c [hunk 23]
```diff
@@ -63,7 +62,7 @@ typedef struct jobqueue {
 	job *rear;               		/* pointer to rear  of queue */
 	bsem *has_jobs;          		/* flag as binary semaphore  */
 	int len;                 		/* number of jobs in queue   */
-	uint64_t max_queued_queries;		/* limit on number of queries */
+	uint64_t cap;		               /* capacity of the queue     */
```
Comment (swilly22) [inline src/util/thpool/thpool.c:65]: make sure all comments starts on the same column number

File: tests/unit/test_thread_pools.cpp [hunk 24]
```diff
@@ -30,7 +29,6 @@ class ThreadPoolsTest: public ::testing::Test {
 	static void SetUpTestCase() {
 		Alloc_Reset();
 		Config_Init(NULL, NULL, 0);
```
Comment (swilly22) [inline tests/unit/test_thread_pools.cpp]: Will this test pass if we remove the call to `Config_Init(NULL, NULL, 0);` ?

**Review by swilly22 — COMMENTED** [review]
Looking really good!

File: tests/unit/test_thread_pools.cpp (hunk 24 above)
Comment (OfirMos) [reply tests/unit/test_thread_pools.cpp]: I still need to check this

File: src/util/rmalloc.c [hunk 25]
```diff
@@ -1,5 +1,91 @@
...
+}
+
+void rm_free_with_capacity(void *ptr) {
+	rm_n_alloced -= (int64_t)malloc_usable_size(ptr);
+	RedisModule_Free(ptr);
+}
+
+void *rm_calloc_with_capacity(size_t nmemb, size_t size) {
+	__alloc(nmemb*size, RedisModule_Calloc, nmemb, size);
+}
```
Comment (swilly22) [inline src/util/rmalloc.c:53]: call `RedisModule_Realloc_Free`

File: src/util/rmalloc.c [hunk 26]
```diff
@@ -1,5 +1,91 @@
...
+
+char *rm_strdup_with_capacity(const char *str) {
+	__alloc((int64_t)malloc_usable_size(str), RedisModule_Strdup, str);
+}
+
+// called when mem_capacity changed
+void rm_set_mem_capacity(int64_t cap) {
+	bool before_limit = (mem_capacity > 0);
+	bool now_limit = (cap > 0);
+	rm_n_alloced = 0;
```
Comment (swilly22) [inline src/util/rmalloc.c:63]: Perform only when applying a capped allocator, add comment explaining why (apply to already executing thread(s))

File: src/util/rmalloc.c [hunk 27]
```diff
@@ -1,5 +1,91 @@
...
+	mem_capacity = cap;
+	asm volatile("" ::: "memory"); // mem barrier cause mem_capacity should be updated before the function pointers
```
Comment (swilly22) [inline src/util/rmalloc.c:65]: Feels a bit too much, we've never encountered issues with instructions reordering

File: src/util/rmalloc.c [hunk 28]
```diff
@@ -1,5 +1,91 @@
...
+		RedisModule_Alloc_Orig = RedisModule_Alloc;
+		RedisModule_Alloc = rm_alloc_with_capacity;
+		RedisModule_Realloc_Orig = RedisModule_Realloc;
+		RedisModule_Realloc = rm_realloc_with_capacity;
+		RedisModule_Free_Orig = RedisModule_Free;
+		RedisModule_Free = rm_free_with_capacity;
+		RedisModule_Calloc_Orig = RedisModule_Calloc;
+		RedisModule_Calloc = rm_calloc_with_capacity;
+		RedisModule_Strdup_Orig = RedisModule_Strdup;
+		RedisModule_Strdup = rm_strdup_with_capacity;
```
Comment (swilly22) [inline src/util/rmalloc.c:76]:
```
if(now_limit && !before_limit) {
// Save original function pointers
		RedisModule_Alloc_Orig     = RedisModule_Alloc;
		RedisModule_Realloc_Orig = RedisModule_Realloc;
		RedisModule_Free_Orig       = RedisModule_Free;
		RedisModule_Calloc_Orig    = RedisModule_Calloc;
		RedisModule_Strdup_Orig   = RedisModule_Strdup;

// Overwrite function pointers to invoke capped allocators
		RedisModule_Realloc = rm_realloc_with_capacity;
		RedisModule_Alloc = rm_alloc_with_capacity;
		RedisModule_Free = rm_free_with_capacity;
		RedisModule_Calloc = rm_calloc_with_capacity;
		RedisModule_Strdup = rm_strdup_with_capacity;
```

File: src/util/rmalloc.c [hunk 29]
```diff
@@ -1,5 +1,91 @@
...
+	} else if(!now_limit && before_limit) {
+		// return all pointers to original functions
```
Comment (swilly22) [inline src/util/rmalloc.c:78]: `// restore all function pointers to their original values`

File: src/util/rmalloc.c [hunk 30]
```diff
@@ -1,5 +1,91 @@
...
+
+void *rm_calloc_with_capacity(size_t nmemb, size_t size) {
+	__alloc(nmemb*size, RedisModule_Calloc, nmemb, size);
+}
+
+char *rm_strdup_with_capacity(const char *str) {
+	__alloc((int64_t)malloc_usable_size(str), RedisModule_Strdup, str);
+}
+
+// called when mem_capacity changed
```
Comment (swilly22) [inline src/util/rmalloc.c:59]: extend comment, explaining the actions taken by it, mentioned the timings

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/arithmetic/algebraic_expression/utils.c [hunk 31]
```diff
@@ -1,5 +1,5 @@
 #include "utils.h"
-#include "../../config.h"
+#include "../../configuration/config.h"
```
Comment (swilly22) [inline src/arithmetic/algebraic_expression/utils.c:2]: sort lines by length

File: src/arithmetic/arithmetic_expression_construct.c [hunk 32]
```diff
@@ -8,7 +8,7 @@
 #include "../RG.h"
 #include "funcs.h"
 #include "../errors.h"
-#include "../config.h"
+#include "../configuration/config.h"
```
Comment (swilly22) [inline src/arithmetic/arithmetic_expression_construct.c:11]: sort lines by length

File: src/arithmetic/path_funcs/path_funcs.c [hunk 33]
```diff
@@ -6,7 +6,7 @@
 
 #include "path_funcs.h"
 #include "../func_desc.h"
-#include "../../config.h"
+#include "../../configuration/config.h"
```
Comment (swilly22) [inline src/arithmetic/path_funcs/path_funcs.c:9]: sort lines by length

File: src/commands/cmd_dispatcher.c [hunk 34]
```diff
@@ -6,7 +6,7 @@
 
 #include "RG.h"
 #include "commands.h"
-#include "../config.h"
+#include "../configuration/config.h"
```
Comment (swilly22) [inline src/commands/cmd_dispatcher.c:9]: sort lines by length

File: src/execution_plan/ops/op_results.c [hunk 35]
```diff
@@ -6,7 +6,7 @@
 
 #include "op_results.h"
 #include "RG.h"
-#include "config.h"
+#include "configuration/config.h"
```
Comment (swilly22) [inline src/execution_plan/ops/op_results.c:9]: sort lines by length

File: src/execution_plan/optimizations/traverse_order.c [hunk 36]
```diff
@@ -4,7 +4,7 @@
 * This file is available under the Redis Labs Source Available License Agreement
 */
 
-#include "../../config.h"
+#include "../../configuration/config.h"
```
Comment (swilly22) [inline src/execution_plan/optimizations/traverse_order.c:7]: sort lines by length

File: src/graph/graph.c [hunk 37]
```diff
@@ -6,7 +6,7 @@
 
 #include "graph.h"
 #include "RG.h"
-#include "config.h"
+#include "configuration/config.h"
```
Comment (swilly22) [inline src/graph/graph.c:9]: sort lines by length

File: src/module.c [hunk 38]
```diff
@@ -9,7 +9,7 @@
 #include "redismodule.h"
 #include "debug.h"
 #include "errors.h"
-#include "config.h"
+#include "configuration/config.h"
```
Comment (swilly22) [inline src/module.c:12]: sort lines by length

File: src/module_event_handlers.c [hunk 39]
```diff
@@ -11,7 +11,7 @@
 #include "graph/graphcontext.h"
 #include "serializers/graphcontext_type.h"
 #include "serializers/graphmeta_type.h"
-#include "config.h"
+#include "configuration/config.h"
```
Comment (swilly22) [inline src/module_event_handlers.c:14]: sort lines by length

File: src/procedures/proc_bfs.c [hunk 40]
```diff
@@ -7,7 +7,7 @@
 #include "proc_bfs.h"
 #include "../RG.h"
 #include "../value.h"
-#include "../config.h"
+#include "../configuration/config.h"
```
Comment (swilly22) [inline src/procedures/proc_bfs.c:10]: sort lines by length

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/util/rmalloc.c [hunk 41]
```diff
@@ -1,4 +1,131 @@
...
+static inline void _nmalloc_increment(int64_t n_bytes) {
+	n_alloced += n_bytes;
+	// check if capacity exceeded
+	if(unlikely(n_alloced > mem_capacity)) {
+		// set n_alloced to MIN to avoid further out of memory exceptions
+		// TODO: consider switching to double -inf
+		n_alloced = INT64_MIN;
+		
+		// throw exception cause memory limit exceeded
+    	ErrorCtx_SetError("Querie's mem consumption exceeded capacity");
```
Comment (jeffreylovitz) [inline src/util/rmalloc.c:59]: Indentation, typo - `Querie` -> `Query`

File: tests/flow/test_query_mem_limit.py [hunk 42]
```diff
@@ -109,8 +113,35 @@ def test_04_overflow_with_limit(self):
...
 
         # expecting out of memory error
         self.env.assertTrue(error_encountered)
 
+    def test_05_test_mixed_queries(self):
+        global error_encountered
+        global n_failed
+        error_encountered = False
+        n_failed = 0
+        n_men_hog_cmds = 10
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:126]: order lines by length

File: tests/flow/test_query_mem_limit.py [hunk 43]
```diff
@@ -109,8 +113,35 @@ def test_04_overflow_with_limit(self):
...
+
+        # set query memory limit to 1MB
+        limit = 1024*1024
+        self.conn.execute_command("GRAPH.CONFIG", "SET", "QUERY_MEM_CAPACITY", limit)
+
+        for i in range(n_men_hog_cmds):
+            n_successfull_queries = random.randint(1, 10)
+            for j in range(n_successfull_queries):
+                self.stress_server(MEM_THRIFTY_QUERY)
+            self.stress_server(MEM_HOG_QUERY)
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:142]:
```
for i in range(total_query_count):
    r = random.randint(0, 100):
    q = MEM_THRIFTY_QUERY
    if r <= total_query_count * 0.1:   #10%
        q = MEM_HOG_QUERY
        expected_error_count += 1
    self.stress_server(q)
```

**Review by swilly22 — CHANGES_REQUESTED** [review]
good thing for that API call...

**Review by swilly22 — APPROVED** [review]
